File: 1472.design-browser-history.ts
/*
 * @lc app=leetcode id=1472 lang=typescript
 *
 * [1472] Design Browser History
 */

// @lc code=start

class HNode {
  url: string;
  prev: HNode | null;
  next: HNode | null;
  constructor(url: string, prev?: HNode | null, next?: HNode | null) {
    this.url = url;
    this.prev = prev === undefined ? null : prev;
    this.next = next === undefined ? null : next;
  }
}

class BrowserHistory {
  homepage: string;
  current: HNode;
  constructor(homepage: string, current = new HNode(homepage)) {
    this.homepage = homepage;
    this.current = current;
  }

  visit(url: string): void {
    let newNode = new HNode(url);
    this.current.next = newNode;
    newNode.prev = this.current;
    this.current = newNode;
  }

  back(steps: number): string {
    let count = 0;
    while (count !== steps) {
      count++;
      if (this.current.prev === null) return this.current.url;
      this.current = this.current.prev;
      if (count === steps) {
        return this.current.url;
      }
    }
    return this.current.url;
  }

  forward(steps: number): string {
    let count = 0;
    while (count !== steps) {
      count++;
      if (this.current.next === null) return this.current.url;
      this.current = this.current.next;
      if (count === steps) {
        return this.current.url;
      }
    }
    return this.current.url;
  }
}

/**
 * Your BrowserHistory object will be instantiated and called as such:
 * var obj = new BrowserHistory(homepage)
 * obj.visit(url)
 * var param_2 = obj.back(steps)
 * var param_3 = obj.forward(steps)
 */
// @lc code=end

File: 1472.design-browser-history.py
#
# @lc app=leetcode id=1472 lang=python3
#
# [1472] Design Browser History
#

# @lc code=start

class BrowserNode:
    def __init__(self, url: str) -> None:
        self.prev = None
        self.next = None
        self.url = url


class BrowserHistory:

    def __init__(self, homepage: str):
        self.head = BrowserNode(homepage)
        self.curr = self.head

    def visit(self, url: str) -> None:
        node = BrowserNode(url)
        self.curr.next = node
        node.prev = self.curr
        self.curr = node

    def back(self, steps: int) -> str:
        while self.curr.prev and steps:
            self.curr = self.curr.prev
            steps -= 1
        return self.curr.url

    def forward(self, steps: int) -> str:
        while self.curr.next and steps:
            self.curr = self.curr.next
            steps -= 1
        return self.curr.url


# Your BrowserHistory object will be instantiated and called as such:
# obj = BrowserHistory(homepage)
# obj.visit(url)
# param_2 = obj.back(steps)
# param_3 = obj.forward(steps)
# @lc code=end
